Cranelift: Generalize `(x << k) >> k` optimization

jameysharp opened the pull request.

jameysharp commented on cranelift/codegen/src/opts/algebraic.isle:
```diff
@@ -198,6 +196,30 @@
         (band ty x (iconst ty mask))))
 
+;; For unsigned shifts, `(x << k) >> k` is the same as masking out the top
+;; `k` bits. A similar rule is valid for vectors but this `iconst` mask only
+;; works for scalar integers.
+(rule (simplify (ushr (fits_in_64 (ty_int ty))
+                      (ishl ty x (iconst _ k))
+                      (iconst _ k)))
+      (band ty x (iconst ty (imm64_ushr ty (imm64 (ty_mask ty)) k))))
```
I'm a little confused about the requirements for `iconst`. Its documentation says it can create "an integer vector where all the lanes have the same value", which probably should work here. Do I just pass the lane type instead of the original `ty` in some of these places? `ty_mask`, at least, and also maybe `imm64_ushr` I guess?

cfallin replied:
Hmm -- that feature is somewhat surprising to me! I don't think we actually have lowerings that do that (and in fact I think we have another opcode for "vector broadcast" or whatever the proper name for that operation is)?

bjorn3 replied:
> and in fact I think we have another opcode for "vector broadcast" or whatever the proper name for that operation is

It's called `splat`.

cfallin commented on cranelift/codegen/src/opts/algebraic.isle:
```diff
@@ -198,6 +196,30 @@
+      ; This if-let is infallible, used for its let-binding effect.
+      (if-let narrow_mask (ty_mask narrow))
+      ; This if-let checks that `narrow_mask` has a subset of the bits that `mask` does.
+      (if-let narrow_mask (u64_and mask narrow_mask))
```
(After talking with Jamey offline about this) it seems this is actually an equality constraint, counterintuitively! The `narrow_mask` is already defined and we start with the existing binding environment, not a fresh one, when building the AST for the LHS pattern; so we get the behavior one would for a second use of a variable in a pattern.

Here we should write an explicit equality (e.g. `(if-let $true (u64_eq ...))`) I think, and separately think about the above binding behavior...

cfallin commented on cranelift/codegen/src/opts/algebraic.isle:
```diff
@@ -198,6 +196,30 @@
+
+;; Masking out the sign-extended bits of an `sextend` turns it into a `uextend`.
+(rule (simplify (band wide (sextend _ x @ (value_type narrow)) (iconst _ (u64_from_imm64 mask))))
+      (if-let mask (ty_mask narrow))
```
(same here as above)

cfallin commented on cranelift/codegen/src/opts/algebraic.isle:
```diff
@@ -204,6 +200,17 @@
+;; `n+k` bits. In the special case where `x` is the result of either
+;; `sextend` or `uextend` from `n` bits to `n+k` bits, we can implement this
+;; using `sextend`.
+(rule (simplify (sshr wide (ishl _ (uextend _ x @ (value_type narrow)) (iconst _ shift)) (iconst _ shift)))
```
This would be a little easier to parse if it were split across multiple lines, IMHO (just a preference!).

cfallin commented on cranelift/codegen/src/opts/algebraic.isle:
```diff
@@ -204,6 +200,17 @@
+;; `sextend` or `uextend` from `n` bits to `n+k` bits, we can implement this
+;; using `sextend`.
+(rule (simplify (sshr wide (ishl _ (uextend _ x @ (value_type narrow)) (iconst _ shift)) (iconst _ shift)))
+      (if-let shift (imm64 (u64_sub (ty_bits_u64 wide) (ty_bits_u64 narrow))))
```
(same comment here re: equality test)

cfallin commented on cranelift/codegen/src/opts/algebraic.isle:
```diff
@@ -204,6 +200,17 @@
+(rule (simplify (sshr wide (ishl _ (uextend _ x @ (value_type narrow)) (iconst _ shift)) (iconst _ shift)))
+      (if-let shift (imm64 (u64_sub (ty_bits_u64 wide) (ty_bits_u64 narrow))))
+      (sextend wide x))
+(rule (simplify (sshr wide (ishl _ x @ (sextend _ (value_type narrow)) (iconst _ shift)) (iconst _ shift)))
```
Just for clarity, even though it's strictly redundant given CLIF typechecking, could we write `wide` for the wildcards for the type params to `ishl` and `sextend` here? (General comment across the rules, and we may not have been consistent elsewhere with this; but especially for rules operating across multiple types, it makes it much easier to reason about!)

cfallin commented on cranelift/codegen/src/opts/algebraic.isle:
```diff
@@ -204,6 +200,17 @@
+      (if-let shift (imm64 (u64_sub (ty_bits_u64 wide) (ty_bits_u64 narrow))))
+      (sextend wide x))
+(rule (simplify (sshr wide (ishl _ x @ (sextend _ (value_type narrow)) (iconst _ shift)) (iconst _ shift)))
+      (if-let shift (imm64 (u64_sub (ty_bits_u64 wide) (ty_bits_u64 narrow))))
```
(same comment re: equality)

cfallin reviewed (approved):
LGTM with comments; thanks!

jameysharp commented:
Hey @cfallin, I think I've addressed all your review comments, but I also realized that I could generalize the rules even further, and added filetests. Could you give this another look? I'm also going to fuzz it for a bit.

jameysharp commented:
I fuzzed this with cranelift-fuzzgen on 1 million inputs over about 3 hours without finding any bugs, and I'd expect that fuzzer to be pretty good at finding bugs in these rules, so I feel pretty confident in this.

cfallin reviewed (approved):
LGTM!
